### PR TITLE
sql: serialize UDTs in expressions in a stable way

### LINKED ISSUE #49379: sql: serialized expressions containing enums should use the physical representation
We serialize expressions in a variety of places in order to parse them back and use, like in expressions for default and computed columns, or in check constraints. In order to be resilient to operations that rename enum members, our serialized enums should contain the unchanging physical representation of the enum, rather than its logical string representation.

### PR DESCRIPTION
**Pull request by rohany**
Fixes #49379.

This PR ensures that serialized expressions stored durably in table
descriptors are serialized in a format that is stable across changes to
user defined types present in those expressions. An effect of this
change is that these expressions must be reparsed and formatted in a
human readable way before display in statements like `SHOW CREATE
TABLE`. 

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
This still depends on some other things getting in, but I wanted to get some preliminary reviews.

**Comment by rohany** [comment]
OK, this seems to mostly work, barring some more formatting related test failures. I'd like some preliminary reviews before digging in and fixing these failures.

File: pkg/sql/sqlbase/structured.go [hunk 1]
```diff
@@ -829,6 +829,83 @@ func (desc *TableDescriptor) AllActiveAndInactiveChecks() []*TableDescriptor_Che
 	return checks
 }
 
+// ReplaceColumnVarsInExprWithDummies replaces the occurrences of column names in an expression with
+// dummies containing their type, so that they may be typechecked. It returns
+// this new expression tree alongside a set containing the ColumnID of each
+// column seen in the expression.
+func (desc *TableDescriptor) ReplaceColumnVarsInExprWithDummies(
```
Comment (otan) [inline pkg/sql/sqlbase/structured.go:836]: i think `ReplaceColumnVarsInExprWithTypeDummy` or something with `Type` in the name. knowing what "dummy" it is is helpful.

File: pkg/sql/sqlbase/structured.go [hunk 2]
```diff
@@ -3893,6 +3970,103 @@ func (desc *ColumnDescriptor) SQLString() string {
 	return f.CloseAndGetString()
 }
 
+// FormatForDisplay formats a column descriptor as a SQL string, and displays
+// user defined types in serialized expressions with human friendly formatting.
+// TODO (rohany): Maybe this should go into a new function?
+// TODO (rohany): This should have a better name?
```
Comment (otan) [inline pkg/sql/sqlbase/structured.go:3976]: `NodeFormat`? :P (does this make it fit into one of the existing related interfaces?)
wdym by "new function"?

File: pkg/sql/sqlbase/structured.go [hunk 3]
```diff
@@ -829,6 +829,83 @@ func (desc *TableDescriptor) AllActiveAndInactiveChecks() []*TableDescriptor_Che
...
 }
 
+// ReplaceColumnVarsInExprWithDummies replaces the occurrences of column names in an expression with
+// dummies containing their type, so that they may be typechecked. It returns
+// this new expression tree alongside a set containing the ColumnID of each
+// column seen in the expression.
+func (desc *TableDescriptor) ReplaceColumnVarsInExprWithDummies(
+	rootExpr tree.Expr,
+) (tree.Expr, map[ColumnID]struct{}, error) {
+	colIDs := make(map[ColumnID]struct{})
```
Comment (otan) [inline pkg/sql/sqlbase/structured.go:839]: should we be using `util.FastIntSet` here?

File: pkg/sql/sqlbase/table.go [hunk 4]
```diff
@@ -183,6 +183,7 @@ func MakeColumnDefDescs(
 		if typedExpr != tree.DNull {
 			d.DefaultExpr.Expr = typedExpr
 			s := tree.Serialize(d.DefaultExpr.Expr)
+			fmt.Println("serialized expr to", s)
```
Comment (otan) [inline pkg/sql/sqlbase/table.go:186]: woops?

File: pkg/sql/sqlbase/structured.go (hunk 1 above)
Comment (otan) [reply pkg/sql/sqlbase/structured.go:836]: Also noting that "Dummy" suggests to me that they are unusable, but seem important for type checking phase. also pondering on names here a little bit. `TypePlaceholder` would be my ideal if placeholder wasn't already inferring something else.

**Review by otan — COMMENTED** [review]
seems sane to me if we're doing this by storing names. does this all work if a name changes? (maybe worth a `rename column` related test?)

i'm wondering whether storing names makes sense. why not descriptor ids? does this overlap with richard's work? what's the future look like?

File: pkg/sql/sqlbase/structured.go (hunk 1 above)
Comment (rohany) [reply pkg/sql/sqlbase/structured.go:836]: My intention with "usable" here means that you shouldn't try to `eval` this. I was thinking placeholder, but that seems to overload with SQL placeholders.

File: pkg/sql/sqlbase/structured.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/sqlbase/structured.go:3976]: There is already a `tree.Serialize` -- we have a precedent for pulling formatting with common flags into functions, and I wanted to make it clear when to use this one vs others.

**Comment by rohany** [comment]
> seems sane to me if we're doing this by storing names. does this all work if a name changes? (maybe worth a rename column related test?)

We don't yet have the ability to rename types :(

> i'm wondering whether storing names makes sense. why not descriptor ids? does this overlap with richard's work? what's the future look like?

I'm confused -- we are storing type descriptor ID's here, and stable representations for enum datums. I don't think it directly overlaps with richards work, but i think in general we want to be moving towards storing IDs in these serialized representations.

**Comment by otan** [comment]
> I'm confused -- we are storing type descriptor ID's here, and stable representations for enum datums. I don't think it directly overlaps with richards work, but i think in general we want to be moving towards storing IDs in these serialized representations.

we're storing Ids in the defaultexpr?

**Comment by rohany** [comment]
We store references to columns as column names. Maybe this should be changed in the future. This PR only affects UDT's

**Comment by rohany** [comment]
Note: depending on this this PR interacts timewise with the partial index work, the same treatment needs to be done for these partial index expressions.

**Comment by rohany** [comment]
I spent some time looking through all of the test failures, and they seem to all be caused by the same thing: 
When preparing a check or computed expression, we would substitute placeholders into the variables of the expression and type check it, but we would serialize _the input untyped expression_. This causes none of the type assertions to get serialized, which seems like a problem. This PR now serializes only typed expressions, so now all the places in the code with them are slightly different.

File: pkg/sql/sqlbase/structured.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/sqlbase/structured.go:3981]: Woops, I read this comment wrong. The new function comment is stale. Perhaps `NodeFormat`!

**Comment by rohany** [comment]
Opened https://github.com/golang/protobuf/issues/1139 for the proto bug I ran into.

File: pkg/sql/sqlbase/structured.go (hunk 3 above)
Comment (rohany) [reply pkg/sql/sqlbase/structured.go]: Yes, done.

**Review by jordanlewis — COMMENTED** [review]
We need to be 100% sure that there's no way this could backfire and leave us with a problem during upgrades from 20.1 to 20.2. Have you thought through this carefully? I can't think of any problems with this, but it's worth checking. A lightweight way to test (without real cross-version roachtests) would be to hardcode a descriptor that has a CHECK constraint serialized in the old way, and force the new code to read, use and modify it...


One concern (unfounded? not sure) is that some ORMs/schema frameworks (cc @rafiss - any ideas here?) check whether they finished a migration by reading the table schema from disk and checking it against the schema that a person originally used in their migration. If the CHECK doesn't roundtrip anymore, since we add these annotations, could that break framework?

Pretty much :lgtm: though!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49565#-:-M8mlDzj-9_g49ez4YuC:ba5dw8z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, @otan, and @rohany)

___
*[pkg/sql/sqlbase/structured.go, line 4025 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49565#-M8mnY6_1i8yR0cuwPJF:-M8mnY6a04-d27b6ZA5h:bl12qdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/acbd408d925d4772e3c97dbd8319f8d926d53b47/pkg/sql/sqlbase/structured.go#L4025)):*
> ```Go
> // SanitizeSerializedExprForDisplay performs this logic, but only returns a
> // tree.Expr to be clear that these returned expressions are not safe to Eval.
> func (desc *TableDescriptor) SanitizeSerializedExprForDisplay(
> ```

nit: Sanitize doesn't seem like that right word. Sanitization to me indicates stripping PII or things of that nature. How about DeserializeTableDescExpr or something of that nature?

Speaking of this, would it makes sense to move this to @mgartner's new `schemaexpr` package, or is that crossing too many wires?

**Comment by rohany** [comment]
> We need to be 100% sure that there's no way this could backfire and leave us with a problem during upgrades from 20.1 to 20.2. Have you thought through this carefully? I can't think of any problems with this, but it's worth checking. A lightweight way to test (without real cross-version roachtests) would be to hardcode a descriptor that has a CHECK constraint serialized in the old way, and force the new code to read, use and modify it...

I thought about it some, and I don't _forsee_ any problems -- all we are doing is adding type annotations, and we don't expect to find them there when we read. Hey, mixed version roachtests are easy (tm) now, maybe I should write one -- meddling with descriptors and then getting SQL to use them isn't fun.

I'm not sure its expected for checks to round trip -- see
```
rohany=# create table checks (x int, y int, check (x > y + 1.0));
CREATE TABLE
rohany=# \d checks;
               Table "public.checks"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 x      | integer |           |          |
 y      | integer |           |          |
Check constraints:
    "checks_check" CHECK (x::numeric > (y::numeric + 1.0))
```
But i think this is a valid concern -- how do ORM's check this? Storing the input string doesn't really work, as we are back to the same problem.

> Speaking of this, would it makes sense to move this to @mgartner's new schemaexpr package, or is that crossing too many wires?

Yeah, that makes sense.

**Comment by rohany** [comment]
> But i think this is a valid concern -- how do ORM's check this? Storing the input string doesn't really work, as we are back to the same problem.

probably by the name of the check constraint lol

**Comment by knz** [comment]
There's a pg catalog table that displays the text of the constraint that's how an orm checks the definition.

**Comment by rohany** [comment]
> There's a pg catalog table that displays the text of the constraint that's how an orm checks the definition.

Yeah, the `consrc` column of `pg_constraint`, but that doesn't round trip with the input -- 

```
rohany=# create table checks (x int, y int, check (x > y + 1.0));
rohany=# select consrc from pg_constraint;
....
 ((x)::numeric > ((y)::numeric + 1.0))
```

**Comment by knz** [comment]
I think you were right that the constraint's name is where it's at. The ORM encodes a version into the identifier (by appending a number) and checks that.

I do recommend you try running some ORM test code to be sure here and guide your design.

**Comment by rafiss** [comment]
Kinda related -- seems like Postgres 12 removes `consrc` entirely. It uses a `pg_get_constraintdef()` function instead, but that also doesn't roundtrip the exact constraint text. I think the name is what is used also.

**Comment by rohany** [comment]
I ran the SQL alchemy tests against my branch as a smoke test and didn't see any unexpected failures. I also tried some version upgrades and confirmed that there isn't an effect here. I'm going to go ahead and clean up this PR now for merging.

File: pkg/sql/logictest/testdata/logic_test/alter_table [hunk 5]
```diff
@@ -983,13 +983,13 @@ ALTER TABLE t ADD CHECK (a < 16) NOT VALID
 query TTTTB
 SHOW CONSTRAINTS FROM t
 ----
-t  check_a   CHECK  CHECK  ((a < 100))  true
-t  check_a1  CHECK  CHECK  ((a < 16))  false
+t  check_a   CHECK  CHECK  ((a < 100:::INT8))  true
+t  check_a1  CHECK  CHECK  ((a < 16:::INT8))   false
```
[the remark this answers is not in the document]
Comment (otan) [reply pkg/sql/logictest/testdata/logic_test/alter_table:987]: (asked offline -- this will be handled separately. we have to crawl these and rewrite values)

**Review by otan — APPROVED** [review]

File: pkg/sql/logictest/testdata/logic_test/enums [hunk 6]
```diff
@@ -469,3 +469,119 @@ WHERE
   descriptor_name = 'enum_array' AND column_name = 'x'
 ----
 x  family:ArrayFamily width:0 precision:0 locale:"" visible_type:0 oid:100064 array_contents:<InternalType:<family:EnumFamily width:0 precision:0 locale:"" visible_type:0 oid:100063 time_precision_is_set:false udt_metadata:<stable_type_id:63 stable_array_type_id:64 > > TypeMeta:<> > time_precision_is_set:false udt_metadata:<stable_type_id:64 stable_array_type_id:0 >
+
+# Test tables using enums in DEFAULT expressions.
+statement ok
+CREATE TABLE enum_default (
+  x INT,
+  y greeting DEFAULT 'hello',
+  z BOOL DEFAULT ('hello':::greeting IS OF (greeting, greeting)),
```
[the remark this answers is not in the document]
Comment (rohany) [reply pkg/sql/logictest/testdata/logic_test/enums:478]: Done

**Comment by rohany** [comment]
tftrs!

bors r=otan,jordanlewis

**Comment by rohany** [comment]
bors r=otan,jordanlewis

**Comment by rohany** [comment]
I think there was an infra flake, trying again.

bors r=otan,jordanlewis

**Comment by rohany** [comment]
bors r=otan,jordanlewis

File: pkg/sql/schemaexpr/check_constraint_test.go [hunk 7]
```diff
@@ -65,15 +65,15 @@ func TestCheckConstraintBuilder_Build(t *testing.T) {
 
 		// Allow expressions that result in a bool.
 		{"ck", "a", true, "a", "ck"},
-		{"ck", "b = 0", true, "b = 0", "ck"},
-		{"ck", "a AND b = 0", true, "a AND (b = 0)", "ck"},
+		{"ck", "b = 0", true, "b = 0:::INT8", "ck"},
+		{"ck", "a AND b = 0", true, "a AND (b = 0:::INT8)", "ck"},
```
Comment (mgartner) [inline pkg/sql/schemaexpr/check_constraint_test.go:69]: Why do only some expressions have type assertions?

Comment (rohany) [reply pkg/sql/schemaexpr/check_constraint_test.go:69]: the serializable formatting flags will serialize all "potentially ambiguous" constants with type annotations -- see the `fmtDisambiguateDatumTypes` formatting flag.

File: pkg/sql/sqlbase/structured.go [hunk 8]
```diff
@@ -835,6 +836,85 @@ func (desc *TableDescriptor) AllActiveAndInactiveChecks() []*TableDescriptor_Che
 	return checks
 }
 
+// ReplaceColumnVarsInExprWithDummies replaces the occurrences of column names in an expression with
+// dummies containing their type, so that they may be typechecked. It returns
+// this new expression tree alongside a set containing the ColumnID of each
+// column seen in the expression.
+func (desc *TableDescriptor) ReplaceColumnVarsInExprWithDummies(
```
Comment (mgartner) [inline pkg/sql/sqlbase/structured.go:843]: Why did you move this to `sqlbase` when it's only used in `schemaexpr`?

Comment (rohany) [reply pkg/sql/sqlbase/structured.go:843]: hmm, this seems like an artifact of rebasing my changes onto the new schema expr package. I'd be happy to move it over, or you can do that if you are making some related changes.

File: pkg/sql/schemaexpr/column.go [hunk 9]
```diff
@@ -99,88 +100,66 @@ func iterColDescriptors(
...
-		vBase, ok := expr.(tree.VarName)
-		if !ok {
-			// Not a VarName, don't do anything to this node.
-			return true, expr, nil
-		}
-
-		v, err := vBase.NormalizeVarName()
+// FormatColumnForDisplay formats a column descriptor as a SQL string, and displays
+// user defined types in serialized expressions with human friendly formatting.
+func FormatColumnForDisplay(
```
Comment (mgartner) [inline pkg/sql/schemaexpr/column.go:132]: Can you add unit tests for this?

File: pkg/sql/schemaexpr/column.go [hunk 10]
```diff
@@ -99,88 +100,66 @@ func iterColDescriptors(
...
+// DeserializeTableDescExpr takes in a serialized expression and a table, and
+// returns an expression that has all user defined types resolved for
+// formatting. It is intended to be used when displaying a serialized
+// expression within a TableDescriptor. For example, a DEFAULT expression
+// of a table containing a user defined type t with id 50 would have all
+// references to t replaced with the id 50. In order to display this expression
+// back to an end user, the serialized id 50 needs to be replaced back with t.
+// DeserializeTableDescExpr performs this logic, but only returns a
+// tree.Expr to be clear that these returned expressions are not safe to Eval.
+func DeserializeTableDescExpr(
```
Comment (mgartner) [inline pkg/sql/schemaexpr/column.go:112]: Can you add unit tests for this?

File: pkg/sql/sqlbase/structured.go (hunk 8 above)
Comment (mgartner) [reply pkg/sql/sqlbase/structured.go:843]: i'll go ahead and move it back in #49819

File: pkg/sql/schemaexpr/column.go (hunk 9 above)
Comment (rohany) [reply pkg/sql/schemaexpr/column.go:132]: Tracking in #49847

File: pkg/sql/schemaexpr/column.go (hunk 10 above)
Comment (rohany) [reply pkg/sql/schemaexpr/column.go:112]: Tracking in #49847
